src/Serilog/Rendering/MessageTemplateRenderer.cs

```diff
@@ -0,0 +1,111 @@
+﻿// Copyright 2013-2017 Serilog Contributors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using Serilog.Events;
+using Serilog.Formatting.Json;
+using Serilog.Parsing;
+
+namespace Serilog.Rendering
+{
+    static class MessageTemplateRenderer
```

Comment:
This class is the remainder - formatting of message templates is externalized from the `MessageTemplate` and related token classes here.

I expect there will be a small perf hit because dispatch is done by switching on type, rather than by virtual dispatch.